### PR TITLE
Support hierarchical namespaces in Azure Blob Storage

### PR DESCRIPTION
**Pull request by nalinigans**
Hierarchical namespaces in Azure Data Storage treat folders as objects now whereas with hierarchical namespaces disabled they exist only in the context of file objects. The `adls` support in [azure-storage-cpplite](https://github.com/nalinigans/azure-storage-cpplite) does have some support to handle directories, but for now have workarounds to figure out paths in `is_dir`, `is_file` and `file_size`. The only thing that cannot be handled with the current base support in `azure-storage-cpplite` is with `delete_dir`. All the blobs themselves can be deleted, but the folder in a hierarchical namespace cannot be. This affects consolidation cosmetically, that is after consolidation, we will be left with empty folders as they cannot be deleted.

Also, modified `delete_dir` to use the blob client directly instead of the wrapper to allow for logging in case of errors.

Update: We have now included `adls` as part of building the `azure-storage-cpplite` library. And are using the [delete_directory](https://github.com/nalinigans/azure-storage-cpplite/blob/1d3f1a41e97f6070a6e9784b30366eb3706b2f54/adls/src/adls_client.cpp#L135) functionality from the `adls_client` for a clean implementation of `delete_dir`.  This is especially important with the consolidation functionality where we want to delete all the older fragments that have been consolidated rather than leave a skeletal directory structure lying around.

### CODE REVIEW

File: core/src/storage_manager/storage_azure_blob.cc [hunk 1]
```diff
@@ -245,12 +245,29 @@ int AzureBlob::set_working_dir(const std::string& dir) {
 }
 
 bool AzureBlob::path_exists(const std::string& path) {
-  bool exists = blob_client_wrapper_->blob_exists(container_name_, get_path(path));
-  if (!exists && path[path.size()-1] == '/') {
+  auto blob_property = blob_client_wrapper_->get_blob_property(container_name_, get_path(path));
+  if (blob_property.valid()) {
+    if (blob_property.content_type == "application/octet-stream") {
```
Comment (mlathara) [inline core/src/storage_manager/storage_azure_blob.cc:250]: Can you help me understand this `application/octet-stream` case? When do we hit that?

File: core/src/storage_manager/storage_azure_blob.cc [hunk 2]
```diff
@@ -245,12 +245,29 @@ int AzureBlob::set_working_dir(const std::string& dir) {
...
+      return path.back() != '/';
+    } else if (path.back() == '/') { // Directories in hierarchical namespaces
+      auto metadata = blob_property.metadata;
+      for (auto prop : metadata) {
+        if (prop.first == "hdi_isfolder" && prop.second == "true") {
```
Comment (mlathara) [inline core/src/storage_manager/storage_azure_blob.cc:255]: Nice find. I don't see this documented here: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties Am I looking in the wrong place?

**Review by mlathara — COMMENTED** [review]
Do you think we need to add the adls stuff to tests?

File: core/src/storage_manager/storage_azure_blob.cc (hunk 1 above)
Comment (nalinigans) [reply core/src/storage_manager/storage_azure_blob.cc]: This is for file blobs in general. The content_type is really the mime type for the content, in our case it is actually octet-stream, we don't really care what type it is, so have resorted to using `blob_property.content_type.empty()` instead.

File: core/src/storage_manager/storage_azure_blob.cc (hunk 2 above)
Comment (nalinigans) [reply core/src/storage_manager/storage_azure_blob.cc]: These are from observations for hierarchical namespaces, but I am now using the same logic as in [adls_client.directory_exists()](https://github.com/nalinigans/azure-storage-cpplite/blob/1d3f1a41e97f6070a6e9784b30366eb3706b2f54/adls/src/adls_client.cpp#L156), but adapted to handling both adls and plain blobs.

File: core/src/storage_manager/storage_azure_blob.cc [hunk 3]
```diff
@@ -200,10 +200,6 @@ AzureBlob::AzureBlob(const std::string& home) {
   }
 
   std::shared_ptr<storage_account> account = std::make_shared<storage_account>(azure_account, cred, /* use_https */true, get_blob_endpoint());
-  if (account == nullptr) {
-    throw std::system_error(EIO, std::generic_category(), "Could not create azure storage account=" + azure_account + ". " +
-                            "Try setting environment variables AZURE_STORAGE_KEY or AZURE_STORAGE_SAS_TOKEN before restarting operation");
-  }
```
Comment (mlathara) [inline core/src/storage_manager/storage_azure_blob.cc:206]: Do we have this check elsewhere?

File: .github/scripts/run_dfs_tests.sh [hunk 4]
```diff
@@ -59,13 +86,20 @@ elif [[ $INSTALL_TYPE == gcs ]]; then
...
-    echo "az schema storage buffer test" && $CMAKE_BUILD_DIR/test/test_storage_buffer --test-dir "az://$AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT.blob.core.windows.net/$TEST" &&
-    echo "az schema examples" && time $GITHUB_WORKSPACE/examples/run_examples.sh "az://$AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT.blob.core.windows.net/$TEST"
-  #echo "Listing Azure Container $AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT";hdfs dfs -ls wasbs://$AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT.blob.core.windows.net/; echo "Listing Azure DONE"
-  #echo "wasbs schema utils test" && tiledb_utils_tests "wasbs://$AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT.blob.core.windows.net/github_unit_test" &&
-  #echo "wasbs schema examples" && time  $GITHUB_WORKSPACE/examples/run_examples.sh "wasbs://$AZURE_CONTAINER_NAME@$AZURE_STORAGE_ACCOUNT.blob.core.windows.net/$TEST"
+  CHECK_RESULTS=(-1 -1)
+  run_azure_tests $GITHUB_WORKSPACE/.github/resources/azure/azure_cred.sh 0 & pids[0]=$!
+  TEST=github_test_${RANDOM}_adls run_azure_tests $GITHUB_WORKSPACE/.github/resources/azure/azure_cred_adls.sh 1 & pids[1]=$!
```
Comment (mlathara) [inline .github/scripts/run_dfs_tests.sh:91]: Are we missing a couple of files from the check-in?
```
$GITHUB_WORKSPACE/.github/resources/azure/azure_cred.sh
$GITHUB_WORKSPACE/.github/resources/azure/azure_cred_adls.sh
```

**Review by mlathara — COMMENTED** [review]
Couple of questions

File: core/src/storage_manager/storage_azure_blob.cc (hunk 3 above)
Comment (nalinigans) [reply core/src/storage_manager/storage_azure_blob.cc:206]: It actually does not make sense to check for a nullptr after a make_shared in the previous line. If there were problems while doing a make_shared, an exception is thrown.

File: .github/scripts/run_dfs_tests.sh (hunk 4 above)
Comment (nalinigans) [reply .github/scripts/run_dfs_tests.sh:91]: They are in the new r.tar.gpg and show up when it is untarred.

**Review by mlathara — APPROVED** [review]
lgtm

**Pull request merged**